2021/test/RingImageInfo.java

```diff
@@ -0,0 +1,108 @@
+
+import team25core.RingDetectionTask;
+import team25core.Robot;
+import team25core.StoneDetectionTask;
+
+class RingImageInfo {
+    private double confidence;
+    private double left;
+    private String ringType;
+    private RingDetectionTask.EventKind ringKind;
+    private double type;
+    private double imageMidpoint;
+    private double ringMidpoint;
+    private double delta;
+    private double margin = 100;
+    private double setColor;
+    private double width;
+    private int imageWidth;
+    private boolean inCenter;
+    private double realNumPixelsPerInch;
+    private final int DISTANCE_FROM_WEBCAM_TO_GRABBER =1;
+    private double distance;
+
+    private int numObjectsSeen;
+    private Telemetry.Item currLocationTlm;
+    private Telemetry.Item ringPositionTlm;
+    private Telemetry.Item ringTlm;
+    private Telemetry.Item ringConfidTlm;
+    private Telemetry.Item ringConfidTlm2;
+    private Telemetry.Item ringTypeTlm;
+    private Telemetry.Item ringTypeTlm2;
+    private Telemetry.Item ringMidpointTlm;
+    private Telemetry.Item imageMidpointTlm;
+    private Telemetry.Item loggingTlm;
+    private Telemetry.Item handleEvntTlm;
+    private Telemetry.Item deltaTlm;
+    private Telemetry.Item numObjectsSeenTlm;
+    private Telemetry.Item pathTlm;
+    private Telemetry.Item widthTlm;
+    private Telemetry.Item marginTlm;
+    private Telemetry.Item imageWidthTlm;
+    private Telemetry.Item pixelsPerInchTlm;
+    private Telemetry.Item distanceBtWWebcamAndGrabberTlm;
+    private Robot myRobot;
+    private Telemetry telemetry;
+
+
+    public RingImageInfo(Robot robot) {
+        myRobot = robot;
+        telemetry = myRobot.telemetry;
+        telemetry.setAutoClear(false);
+        //caption: what appears on the phone
+        ringPositionTlm = telemetry.addData("LeftOrigin", "unknown");
+        ringConfidTlm = telemetry.addData("Confidence", "N/A");
+        ringConfidTlm2 = telemetry.addData("Confidence2", "N/A");
+        ringTypeTlm = telemetry.addData("RingType","unknown");
+        ringTypeTlm2 = telemetry.addData("RingType2","unknown");
+        imageMidpointTlm = telemetry.addData("Image_Mdpt", "unknown");
+        ringMidpointTlm = telemetry.addData("Ring Mdpt", "unknown");
+        ringTlm = telemetry.addData("kind", "unknown");
+        deltaTlm = telemetry.addData("delta", "unknown");
+        numObjectsSeenTlm = telemetry.addData("numRings",-1);
+        pathTlm = telemetry.addData("AllianceClr", "unknown");
+        widthTlm = telemetry.addData("ringWidth", "unknown");
+        imageWidthTlm = telemetry.addData("imageWidth", -1);
+        marginTlm = telemetry.addData("margin" , "unknown");
+        pixelsPerInchTlm = telemetry.addData("pixelsPerInch", "unknown");
+        distanceBtWWebcamAndGrabberTlm = telemetry.addData("distance BtW Webcam and Grabber","unknown");
+        currLocationTlm = telemetry.addData("curr location", "in RingImageInfo constructor");
```

Comment:
I'd recommend moving all of the telemetry out of the constructor.  That would remove the dependency on Robot, and make the class usable outside the context having to pass it a Robot instance.

I'd add

```
public void displayTelemetry(Telemetry telemetry) { ... }
```

Move all the telemetry statements above into that method, and then call it from the opmode when you want to display telemetry.  (This also has the added benefit of being able to run an opmode and use this class without all the telemetry spamming the DS display (which might be useful later on in the season)).